2-Draft-Accepted/RFCxxxx-PowerShellGet-3.0.md

```diff
@@ -0,0 +1,242 @@
+different resources types that it supports:
+
+`Find-Command`
+`Find-DscResource`
+`Find-Module`
+`Find-RoleCapability`
+`Find-Script`
+
+Instead, these will be abstracted by a `Find-PSResource` cmdlet.
+A `-Type` parameter will accept an array of types to allow filtering.
+
+With support of the generic `PSResource`, this means we can also find and
+install arbitrary nupkgs that may only contain an assembly the user wants to
+use in their script.
+
+If the local cache does not exist, the cmdlet will call `Update-PSResourceCache`
+first and then attempt a search.
+However, this cmdlet does not update the cache if one exists.
+
+### Installing resources
+
+`Install-PSResource` will only work for modules.
+A `-AllowPrerelease` switch allows installing prerelease versions.
+Other types will use `Save-PSResource` (see below).
+A `-Repository` parameter accepts a specific repository name.
+If `-Repository` is not specified and the same resource is found in multiple
+repositories, then the resource is automatically installed quietly from the
+trusted repository with the highest version matching the `-Version` parameter
+(if specified, otherwise newest non-prerelease version unless `-AllowPrerelease`
+is used).
+If there are no trusted repositories matching the query, then the newest version
+fulfilling the query will be prompted to be installed.
+If there are multiple repositories with the same trust level containing the same
+version, the first one is used.
+
+`-AllowUntrusted` can be used to suppress being prompted for untrusted sources.
+`-AllowDifferentPublisher` can be used to suppress being prompted if the publisher
+of the module is different from the currently installed version.
+`-AllowReinstall` can be used to re-install a module.
+
+A `-Quiet` switch will suppress progress information.
+
+### Dependencies and version management
+
+`Install-PSResource` can accept a path to a psd1 file (using `-RequiredModulesFile`)
+or a hashtable (using `-RequiredModules`) where the key is the module name and the
+value is either the required version specified using Nuget version range syntax or
+a hash table where `repository` is set to the URL of the repository and
+`version` contains the [Nuget version range syntax](https://docs.microsoft.com/en-us/nuget/reference/package-versioning#version-ranges-and-wildcards).
+
+```powershell
+Install-PSResource -RequiredModules @{
+  "Configuration" = "[1.3.1,2.0)"
+  "Pester"        = @{
+    version = "[4.4.2,4.7.0]"
+    repository = "https://www.powershellgallery.com"
+  }
+}
+```
+
+### Saving resources
+
+With the removal of PackageManagement, there is still a need to support saving
+arbitrary nupkgs (assemblies) used for scripts.
+
+`Save-PSResource -Type Library` will download nupkgs that have a `lib` folder.
+A `-AllowPrerelease` switch allows saving prerelease versions.
+The dependent native library in `runtimes` matching the current system runtime
+will be copied to the same destination.
+
```

Comment:
I'd also like a parameter for either all runtimes or a list of runtimes so that we can use this to pull binary dependencies for cross-platform and cross-framework modules

Reply:
A `-IncludeAllRuntimes` would make sense.  I think the default for most users is the runtime that applies to their system.

Reply:
If we're unpacking just the binaries for this runtime, are we putting them in a user-specified path (default current?) or in a `lib` subfolder of it, or in a `lib\runtime\` subfolder of it?

If we're keeping them all, how does the answer to that change?

Should this just be an option to keep the whole nupkg file instead? If the developer wants everything, maybe they need the tools folder or whatever else, also?

Reply:
Will clarify.  For the default runtime, it'll be in the root of the specified destination.  For `-IncludeAllRuntimes`, it'll just copy the `runtimes` folder verbatim to the destination.